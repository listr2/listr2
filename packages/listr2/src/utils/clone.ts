import rfdc from 'rfdc'

const clone = rfdc({ circles: true })

/**
 * Deep clones a object in the easiest manner.
 */
// TODO: swap this with structured clone whenever we migrate to node 18
export function cloneObject<T extends Record<PropertyKey, any>>(obj: T): T {
  return clone(obj)
}
