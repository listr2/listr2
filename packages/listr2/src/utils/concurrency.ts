// adapted from https://github.com/fabiospampinato/promise-concurrency-limiter
// with MIT license, since it is again ESM only
// we needed to stop on the first failing case unlike the original
export class Concurrency {
  promise: Promise<void>

  private resolve: () => void
  private reject: (error: Error) => void
  private concurrency: number
  private count: number
  private queue: Set<() => void>

  constructor (options: { concurrency: number }) {
    this.concurrency = options.concurrency
    this.count = 0
    this.queue = new Set()
    this.initPromise()
  }

  public add<T>(fn: () => Promise<T>): Promise<T> {
    if (this.count < this.concurrency) {
      return this.run(fn)
    }

    return new Promise<T>((resolve) => {
      const callback = (): void => resolve(this.run(fn))

      this.queue.add(callback)
    })
  }

  private initPromise (): void {
    const { promise, resolve, reject } = this.withResolvers<void>()

    this.promise = promise
    this.resolve = resolve
    this.reject = reject
  }

  private flush (): void {
    for (const callback of this.queue) {
      if (this.count >= this.concurrency) {
        break
      }

      this.queue.delete(callback)

      callback()
    }
  }

  private run<T>(fn: () => Promise<T>): Promise<T> {
    this.count++

    const promise = fn()

    const cleanup = (): void => {
      this.count--

      this.flush()

      if (this.count === 0) {
        this.resolve()
      }
    }

    promise.then(cleanup, (error) => {
      this.queue.clear()
      this.reject(error)
    })

    return promise
  }

  private withResolvers<T>(): { promise: Promise<T>, resolve: (value: T) => void, reject: (error: Error) => void } {
    let resolve: (value: T) => void
    let reject: (error: Error) => void

    const promise = new Promise<T>((res, rej) => {
      resolve = res
      reject = rej
    })

    return {
      promise,
      resolve: resolve!,
      reject: reject!
    }
  }
}
