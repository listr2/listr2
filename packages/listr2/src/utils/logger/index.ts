export * from './logger.constants'
export type * from './logger.interface'
export * from './logger'
