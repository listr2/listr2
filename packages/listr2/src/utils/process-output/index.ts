export type * from './process-output-buffer.interface'
export * from './process-output-buffer'
export * from './process-output-stream'
export * from './process-output'
export type * from './process-output.interface'
export * from './writable'
