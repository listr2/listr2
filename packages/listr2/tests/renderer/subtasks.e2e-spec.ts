import { Listr, delay } from '@root'
import type { MockProcessOutput, RendererSetup } from '@tests/utils'
import { expectProcessOutputToMatchSnapshot, mockProcessOutput, unmockProcessOutput, RENDERER_SETUP } from '@tests/utils'

describe.each<RendererSetup>(RENDERER_SETUP)('%s renderer: show subtasks', (renderer, rendererOptions) => {
  const output: MockProcessOutput = {} as MockProcessOutput

  process.stdout.isTTY = true

  beforeEach(async() => {
    mockProcessOutput(output)
  })

  afterEach(async() => {
    unmockProcessOutput(output)
    jest.clearAllMocks()
  })

  // JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR
  it('should create a set of subtasks', async() => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: (_, task): Listr =>
            task.newListr([
              {
                title: 'This is a subtask.',
                task: async(): Promise<void> => {}
              },

              {
                title: 'This is a subtask.',
                task: async(): Promise<void> => {}
              }
            ])
        }
      ],
      {
        concurrent: false,
        renderer,
        rendererOptions
      }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR')
  })

  // 3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k
  it('should be able to change concurrency setting in subtask, second subtask should finish first', async() => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: (_, task): Listr =>
            task.newListr(
              [
                {
                  title: 'This is a subtask.',
                  task: async(): Promise<void> => {
                    await delay(10)
                  }
                },

                {
                  title: 'This is an another subtask.',
                  task: async(): Promise<void> => {}
                }
              ],
              { concurrent: true }
            )
        }
      ],
      {
        concurrent: false,
        renderer,
        rendererOptions
      }
    ).run()

    expectProcessOutputToMatchSnapshot(output, '3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k')
  })

  // 12ttDlnth5pr1YuJEnfXm3I2CzRbcFlY
  it('should be able to change exit on error', async() => {
    try {
      await new Listr(
        [
          {
            title: 'This task will execute and not quit on errors.',
            task: (_, task): Listr =>
              task.newListr(
                [
                  {
                    title: 'This is a subtask.',
                    task: async(): Promise<void> => {
                      throw new Error('I have failed [0]')
                    }
                  },
                  {
                    title: 'This is an another subtask.',
                    task: async(): Promise<void> => {
                      throw new Error('I have failed [1]')
                    }
                  },
                  {
                    title: 'This is yet an another subtask.',
                    task: async(_, task): Promise<void> => {
                      task.title = 'I have succeeded.'
                    }
                  }
                ],
                { exitOnError: false }
              )
          },
          {
            title: 'This task will execute.',
            task: (): void => {
              throw new Error('exit')
            }
          }
        ],
        {
          concurrent: false,
          exitOnError: true,
          renderer,
          rendererOptions
        }
      ).run()
    } catch (e: any) {
      expect(e).toBeTruthy()
    }

    expectProcessOutputToMatchSnapshot(output, '12ttDlnth5pr1YuJEnfXm3I2CzRbcFlY')
  })
})
