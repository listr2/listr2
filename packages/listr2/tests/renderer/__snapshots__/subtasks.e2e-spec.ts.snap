// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`default renderer: show subtasks should be able to change concurrency setting in subtask, second subtask should finish first: 3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k-exit 1`] = `[]`;

exports[`default renderer: show subtasks should be able to change concurrency setting in subtask, second subtask should finish first: 3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k-stderr 1`] = `[]`;

exports[`default renderer: show subtasks should be able to change concurrency setting in subtask, second subtask should finish first: 3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k-stdout 1`] = `
[
  "[?25l",
  "‚óº This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
  ‚óº This is a subtask.
",
  "[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚óº This is a subtask.
  ‚óº This is an another subtask.
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚ùØ This is a subtask.
  ‚óº This is an another subtask.
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚ùØ This is a subtask.
  ‚ùØ This is an another subtask.
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚ùØ This is a subtask.
  ‚úî This is an another subtask.
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚úî This is a subtask.
  ‚úî This is an another subtask.
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will execute.
",
  "[2K[1A[2K[G",
  "‚úî This task will execute.
",
  "[?25h",
]
`;

exports[`default renderer: show subtasks should be able to change exit on error: 12ttDlnth5pr1YuJEnfXm3I2CzRbcFlY-exit 1`] = `[]`;

exports[`default renderer: show subtasks should be able to change exit on error: 12ttDlnth5pr1YuJEnfXm3I2CzRbcFlY-stderr 1`] = `[]`;

exports[`default renderer: show subtasks should be able to change exit on error: 12ttDlnth5pr1YuJEnfXm3I2CzRbcFlY-stdout 1`] = `
[
  "[?25l",
  "‚óº This task will execute and not quit on errors.
",
  "[2K[1A[2K[G‚óº This task will execute and not quit on errors.
‚óº This task will execute.
",
  "[2K[1A[2K[1A[2K[G‚ùØ This task will execute and not quit on errors.
‚óº This task will execute.
",
  "[2K[1A[2K[1A[2K[G‚ùØ This task will execute and not quit on errors.
  ‚óº This is a subtask.
‚óº This task will execute.
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute and not quit on errors.
  ‚óº This is a subtask.
  ‚óº This is an another subtask.
‚óº This task will execute.
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute and not quit on errors.
  ‚óº This is a subtask.
  ‚óº This is an another subtask.
  ‚óº This is yet an another subtask.
‚óº This task will execute.
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute and not quit on errors.
  ‚ùØ This is a subtask.
  ‚óº This is an another subtask.
  ‚óº This is yet an another subtask.
‚óº This task will execute.
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute and not quit on errors.
  ‚úñ This is a subtask.
  ‚óº This is an another subtask.
  ‚óº This is yet an another subtask.
‚óº This task will execute.
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute and not quit on errors.
  ‚úñ I have failed [0]
  ‚óº This is an another subtask.
  ‚óº This is yet an another subtask.
‚óº This task will execute.
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute and not quit on errors.
  ‚úñ I have failed [0]
  ‚ùØ This is an another subtask.
  ‚óº This is yet an another subtask.
‚óº This task will execute.
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute and not quit on errors.
  ‚úñ I have failed [0]
  ‚úñ This is an another subtask.
  ‚óº This is yet an another subtask.
‚óº This task will execute.
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute and not quit on errors.
  ‚úñ I have failed [0]
  ‚úñ I have failed [1]
  ‚óº This is yet an another subtask.
‚óº This task will execute.
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute and not quit on errors.
  ‚úñ I have failed [0]
  ‚úñ I have failed [1]
  ‚ùØ This is yet an another subtask.
‚óº This task will execute.
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute and not quit on errors.
  ‚úñ I have failed [0]
  ‚úñ I have failed [1]
  ‚ùØ I have succeeded.
‚óº This task will execute.
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute and not quit on errors.
  ‚úñ I have failed [0]
  ‚úñ I have failed [1]
  ‚úî I have succeeded.
‚óº This task will execute.
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ö† This task will execute and not quit on errors.
  ‚úñ I have failed [0]
  ‚úñ I have failed [1]
  ‚úî I have succeeded.
‚óº This task will execute.
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ö† This task will execute and not quit on errors.
  ‚úñ I have failed [0]
  ‚úñ I have failed [1]
  ‚úî I have succeeded.
‚ùØ This task will execute.
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ö† This task will execute and not quit on errors.
  ‚úñ I have failed [0]
  ‚úñ I have failed [1]
  ‚úî I have succeeded.
‚úñ This task will execute.
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ö† This task will execute and not quit on errors.
  ‚úñ I have failed [0]
  ‚úñ I have failed [1]
  ‚úî I have succeeded.
‚úñ exit
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G",
  "‚ö† This task will execute and not quit on errors.
  ‚úñ I have failed [0]
  ‚úñ I have failed [1]
  ‚úî I have succeeded.
‚úñ exit
",
  "[?25h",
]
`;

exports[`default renderer: show subtasks should create a set of subtasks: JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR-exit 1`] = `[]`;

exports[`default renderer: show subtasks should create a set of subtasks: JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR-stderr 1`] = `[]`;

exports[`default renderer: show subtasks should create a set of subtasks: JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR-stdout 1`] = `
[
  "[?25l",
  "‚óº This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
",
  "[2K[1A[2K[G‚ùØ This task will execute.
  ‚óº This is a subtask.
",
  "[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚óº This is a subtask.
  ‚óº This is a subtask.
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚ùØ This is a subtask.
  ‚óº This is a subtask.
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚úî This is a subtask.
  ‚óº This is a subtask.
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚úî This is a subtask.
  ‚ùØ This is a subtask.
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
  ‚úî This is a subtask.
  ‚úî This is a subtask.
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will execute.
",
  "[2K[1A[2K[G",
  "‚úî This task will execute.
",
  "[?25h",
]
`;

exports[`simple renderer: show subtasks should be able to change concurrency setting in subtask, second subtask should finish first: 3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k-exit 1`] = `[]`;

exports[`simple renderer: show subtasks should be able to change concurrency setting in subtask, second subtask should finish first: 3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k-stderr 1`] = `[]`;

exports[`simple renderer: show subtasks should be able to change concurrency setting in subtask, second subtask should finish first: 3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k-stdout 1`] = `
[
  "‚ùØ This task will execute.
",
  "‚ùØ This is a subtask.
",
  "‚ùØ This is an another subtask.
",
  "‚úî This is an another subtask.
",
  "‚úî This is a subtask.
",
  "‚úî This task will execute.
",
]
`;

exports[`simple renderer: show subtasks should be able to change exit on error: 12ttDlnth5pr1YuJEnfXm3I2CzRbcFlY-exit 1`] = `[]`;

exports[`simple renderer: show subtasks should be able to change exit on error: 12ttDlnth5pr1YuJEnfXm3I2CzRbcFlY-stderr 1`] = `
[
  "‚úñ This is a subtask. [FAILED: I have failed [0]]
",
  "‚úñ This is an another subtask. [FAILED: I have failed [1]]
",
  "‚úñ This task will execute. [FAILED: exit]
",
]
`;

exports[`simple renderer: show subtasks should be able to change exit on error: 12ttDlnth5pr1YuJEnfXm3I2CzRbcFlY-stdout 1`] = `
[
  "‚ùØ This task will execute and not quit on errors.
",
  "‚ùØ This is a subtask.
",
  "‚ùØ This is an another subtask.
",
  "‚ùØ This is yet an another subtask.
",
  "‚úî I have succeeded.
",
  "‚úî This task will execute and not quit on errors.
",
  "‚ùØ This task will execute.
",
]
`;

exports[`simple renderer: show subtasks should create a set of subtasks: JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR-exit 1`] = `[]`;

exports[`simple renderer: show subtasks should create a set of subtasks: JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR-stderr 1`] = `[]`;

exports[`simple renderer: show subtasks should create a set of subtasks: JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR-stdout 1`] = `
[
  "‚ùØ This task will execute.
",
  "‚ùØ This is a subtask.
",
  "‚úî This is a subtask.
",
  "‚ùØ This is a subtask.
",
  "‚úî This is a subtask.
",
  "‚úî This task will execute.
",
]
`;

exports[`test renderer: show subtasks should be able to change concurrency setting in subtask, second subtask should finish first: 3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k-exit 1`] = `[]`;

exports[`test renderer: show subtasks should be able to change concurrency setting in subtask, second subtask should finish first: 3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k-stderr 1`] = `[]`;

exports[`test renderer: show subtasks should be able to change concurrency setting in subtask, second subtask should finish first: 3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"This task will execute.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute."]}}
",
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"This is a subtask.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute.","This is a subtask."]}}
",
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"This is an another subtask.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute.","This is an another subtask."]}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":true,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"This is an another subtask.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute.","This is an another subtask."]}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":true,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"This is a subtask.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute.","This is a subtask."]}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":true,"isSkipped":false,"hasFinalized":true,"hasSubtasks":true,"title":"This task will execute.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute."]}}
",
]
`;

exports[`test renderer: show subtasks should be able to change exit on error: 12ttDlnth5pr1YuJEnfXm3I2CzRbcFlY-exit 1`] = `[]`;

exports[`test renderer: show subtasks should be able to change exit on error: 12ttDlnth5pr1YuJEnfXm3I2CzRbcFlY-stderr 1`] = `
[
  "{"event":"MESSAGE","data":{"error":"I have failed [0]"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"This is a subtask.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":["This task will execute and not quit on errors.","This is a subtask."]}}
",
  "{"event":"MESSAGE","data":{"error":"I have failed [1]"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"This is an another subtask.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":["This task will execute and not quit on errors.","This is an another subtask."]}}
",
  "{"event":"MESSAGE","data":{"error":"exit"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"This task will execute.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":["This task will execute."]}}
",
]
`;

exports[`test renderer: show subtasks should be able to change exit on error: 12ttDlnth5pr1YuJEnfXm3I2CzRbcFlY-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"This task will execute and not quit on errors.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute and not quit on errors."]}}
",
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"This is a subtask.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute and not quit on errors.","This is a subtask."]}}
",
  "{"event":"STATE","data":"FAILED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"This is a subtask.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":["This task will execute and not quit on errors.","This is a subtask."]}}
",
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"This is an another subtask.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute and not quit on errors.","This is an another subtask."]}}
",
  "{"event":"STATE","data":"FAILED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"This is an another subtask.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":["This task will execute and not quit on errors.","This is an another subtask."]}}
",
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"This is yet an another subtask.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute and not quit on errors.","This is yet an another subtask."]}}
",
  "{"event":"TITLE","data":"I have succeeded.","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"I have succeeded.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute and not quit on errors.","This is yet an another subtask."]}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":true,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"I have succeeded.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute and not quit on errors.","This is yet an another subtask."]}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":true,"isSkipped":false,"hasFinalized":true,"hasSubtasks":true,"title":"This task will execute and not quit on errors.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute and not quit on errors."]}}
",
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"This task will execute.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute."]}}
",
  "{"event":"STATE","data":"FAILED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"This task will execute.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":["This task will execute."]}}
",
]
`;

exports[`test renderer: show subtasks should create a set of subtasks: JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR-exit 1`] = `[]`;

exports[`test renderer: show subtasks should create a set of subtasks: JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR-stderr 1`] = `[]`;

exports[`test renderer: show subtasks should create a set of subtasks: JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"This task will execute.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute."]}}
",
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"This is a subtask.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute.","This is a subtask."]}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":true,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"This is a subtask.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute.","This is a subtask."]}}
",
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"This is a subtask.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute.","This is a subtask."]}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":true,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"This is a subtask.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute.","This is a subtask."]}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":true,"isSkipped":false,"hasFinalized":true,"hasSubtasks":true,"title":"This task will execute.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["This task will execute."]}}
",
]
`;

exports[`verbose renderer: show subtasks should be able to change concurrency setting in subtask, second subtask should finish first: 3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k-exit 1`] = `[]`;

exports[`verbose renderer: show subtasks should be able to change concurrency setting in subtask, second subtask should finish first: 3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k-stderr 1`] = `[]`;

exports[`verbose renderer: show subtasks should be able to change concurrency setting in subtask, second subtask should finish first: 3ygUxrTkCDLDfDrXzF1ocWR6626jaL0k-stdout 1`] = `
[
  "[STARTED] This task will execute.
",
  "[STARTED] This is a subtask.
",
  "[STARTED] This is an another subtask.
",
  "[COMPLETED] This is an another subtask.
",
  "[COMPLETED] This is a subtask.
",
  "[COMPLETED] This task will execute.
",
]
`;

exports[`verbose renderer: show subtasks should be able to change exit on error: 12ttDlnth5pr1YuJEnfXm3I2CzRbcFlY-exit 1`] = `[]`;

exports[`verbose renderer: show subtasks should be able to change exit on error: 12ttDlnth5pr1YuJEnfXm3I2CzRbcFlY-stderr 1`] = `
[
  "[FAILED] I have failed [0]
",
  "[FAILED] I have failed [1]
",
  "[FAILED] exit
",
]
`;

exports[`verbose renderer: show subtasks should be able to change exit on error: 12ttDlnth5pr1YuJEnfXm3I2CzRbcFlY-stdout 1`] = `
[
  "[STARTED] This task will execute and not quit on errors.
",
  "[STARTED] This is a subtask.
",
  "[STARTED] This is an another subtask.
",
  "[STARTED] This is yet an another subtask.
",
  "[TITLE] I have succeeded.
",
  "[COMPLETED] I have succeeded.
",
  "[COMPLETED] This task will execute and not quit on errors.
",
  "[STARTED] This task will execute.
",
]
`;

exports[`verbose renderer: show subtasks should create a set of subtasks: JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR-exit 1`] = `[]`;

exports[`verbose renderer: show subtasks should create a set of subtasks: JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR-stderr 1`] = `[]`;

exports[`verbose renderer: show subtasks should create a set of subtasks: JQYWVb3x1scokQK2IzhwA4F0qxWbYzXR-stdout 1`] = `
[
  "[STARTED] This task will execute.
",
  "[STARTED] This is a subtask.
",
  "[COMPLETED] This is a subtask.
",
  "[STARTED] This is a subtask.
",
  "[COMPLETED] This is a subtask.
",
  "[COMPLETED] This task will execute.
",
]
`;
